### PR TITLE
feat(utils/function): add sleep util function

### LINKED ISSUE #47: Create sleep Utility Function
### `sleep`

Pauses execution for a specified number of milliseconds.

**File**:  `src/utils/function/sleep/sleep.ts`
**For Test File**:  `src/utils/function/sleep/sleep.test.ts`

**Parameters**:
- `ms: number`: The number of milliseconds to sleep.

**Returns**:
- `Promise<void>`: A promise that resolves after the specified delay.

**Example**:
```ts
await sleep(1000);  // Pauses for 1 second
```

### PR DESCRIPTION
**Pull request by Faraz-Khan-79996**
This PR adds a sleep utility function that pauses execution for a specified number of milliseconds.

Closes #47

## Summary by CodeRabbit

- **New Features**
	- Introduced several new utility functions: `sleep`, `clamp`, `hasKey`, and `toUpperCase`, enhancing data manipulation and control within the application.
	- Added a custom React hook `usePrevious` to track previous values of variables in functional components.
	- Implemented structured error handling with a new `ErrorHandler` class for clearer error management.

- **Tests**
	- Added comprehensive test suites for the new utility functions and the `sleep` function, ensuring their reliability across various scenarios.

### CODE REVIEW

File: src/utils/function/sleep.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * @param ms - The number of milliseconds to sleep.
+ * @returns A promise that resolves after the specified delay.
+ */
+export const sleep = (ms: number): Promise<void> => {
+    return new Promise(resolve => setTimeout(resolve, ms));
+};
+
+// Example usage:
+async function example() {
```
Comment (bharat-dussa) [inline src/utils/function/sleep.ts:14]: don't add example code

File: src/utils/function/sleep.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// src/utils/function/sleep.ts
+
+/**
+ * Pauses execution for a specified number of milliseconds.
+ *
+ * @param ms - The number of milliseconds to sleep.
+ * @returns A promise that resolves after the specified delay.
+ */
+export const sleep = (ms: number): Promise<void> => {
+    return new Promise(resolve => setTimeout(resolve, ms));
```
Comment (bharat-dussa) [inline src/utils/function/sleep.ts:10]: handle error handling. 

- [ ] should not accept non number
- [ ] should not accept negative number

throw TypeError if it doesn't meet requirements

File: src/utils/function/sleep.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// src/utils/function/sleep.ts
```
Comment (bharat-dussa) [inline src/utils/function/sleep.ts:1]: change PR title, 
it should be like

`feat(utils/function): add sleep util function`

Points to add:

- [ ] file should be imported in `utils/index.ts`
- [ ] add test file for this
- [ ] add types in types/utils/sleep.d.ts

if any help required you can contact at https://discord.gg/qNeQBF2Q

**Review by bharat-dussa — CHANGES_REQUESTED** [review]

File: src/utils/function/sleep.ts (hunk 1 above)
Comment (Faraz-Khan-79996) [reply src/utils/function/sleep.ts]: removed it.

File: src/utils/function/sleep.ts (hunk 2 above)
Comment (Faraz-Khan-79996) [reply src/utils/function/sleep.ts]: Handled errors.

File: src/utils/function/sleep.ts (hunk 3 above)
Comment (Faraz-Khan-79996) [reply src/utils/function/sleep.ts:1]: PR title changed.
File imported in index.ts.
test file added to path.
Type added in file.

File: src/utils/function/sleep.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ * @returns A promise that resolves after the specified delay.
+ * @throws Error if the input is not a valid positive number.
+ */
+export const sleep = (ms: number): Promise<void> => {
+    if (typeof ms !== 'number' || isNaN(ms) || ms < 0) {
+        throw new Error("Invalid input: ms should be a non-negative number.");
+    }
+
+    return new Promise(resolve => setTimeout(resolve, ms));
+};
```
[the remark this answers is not in the document]
Comment (Faraz-Khan-79996) [reply src/utils/function/sleep.ts]: did what was asked

File: src/utils/function/sleep/sleep.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * Pauses execution for a specified number of milliseconds.
+ *
+ * @param ms - The number of milliseconds to sleep.
+ * @returns A promise that resolves after the specified delay.
+ * @throws Error if the input is not a valid positive number.
+ */
+export const sleep = (ms: number): Promise<void> => {
+    if (typeof ms !== 'number' || isNaN(ms) || ms < 0) {
+        throw new Error("Invalid input: ms should be a non-negative number.");
+    }
```
[the remark this answers is not in the document]
Comment (Faraz-Khan-79996) [reply src/utils/function/sleep/sleep.ts]: did what you asked.

File: src/utils/function/sleep/sleep.test.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { sleep } from './sleep';
+
+test('sleep function pauses for specified milliseconds', async () => {
```
Comment (bharat-dussa) [inline src/utils/function/sleep/sleep.test.ts:3]:
```suggestion
 describe("sleep function", () => {
   it("sleep function pauses for specified milliseconds", ...)
 })
```

File: src/utils/function/sleep/sleep.test.ts
Comment (bharat-dussa) [inline src/utils/function/sleep/sleep.test.ts:1]: add tests for multiple secenarios 
1. should resolve after the specified time
2. should resolve immediately when given 0
3. should throw an error when given a negative number
4. should throw an error when given NaN
5. should throw an error when given a non-number type
6. should resolve correctly with a large number
7. should allow multiple asynchronous calls
8. should handle multiple subsequent calls
9. should handle multiple simultaneous calls correctly

File: src/utils/function/sleep/sleep.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Pauses execution for a specified number of milliseconds.
+ *
+ * @param ms - The number of milliseconds to sleep.
+ * @returns A promise that resolves after the specified delay.
+ * @throws Error if the input is not a valid positive number.
+ */
+export const sleep = (ms: number): Promise<void> => {
+    if (typeof ms !== 'number' || Number.isNaN(ms) || ms < 0) {
```
Comment (bharat-dussa) [inline src/utils/function/sleep/sleep.ts:11]: Nice Doc! and function implementation

Comment (bharat-dussa) [reply src/utils/function/sleep/sleep.ts:11]: However there are few points to improve for tests, please resolve those

**Review by bharat-dussa — CHANGES_REQUESTED** [review]

**Comment by bharat-dussa** [comment]
@Faraz-Khan-79996 Thank you for your contribution, made granular conditions for explicit errors 

Still count in HacktoberFest 
So Rest Assured!

Due to short time! I am merging this, Please feel to raise or resolve other issues. Any idea connect to discord.

Keep typing...

**Pull request merged**